### PR TITLE
feat: Configuring the FQDN GC interval / Zombie timeout

### LINKED ISSUE #13687: Support configuring the FQDN GC interval / Zombie timeout
This line defines the FQDN GC interval:

https://github.com/cilium/cilium/blob/351a3764299a8466ea45c9fc05a0dfc525ce2f68/daemon/cmd/fqdn.go#L162

Which is subsequently used to determine the timeout for "zombie" FQDN mappings:

https://github.com/cilium/cilium/blob/351a3764299a8466ea45c9fc05a0dfc525ce2f68/daemon/cmd/fqdn.go#L170-L175

This determines how long that connections should be idle for after the DNS TTL expires.

If we made the latter TTL configurable, then users could tweak the FQDN mapping expiry to the TCP keepalive behaviour of their applications (assuming the application continues to use DNS names beyond the TTL of the FQDN).

Given the comment above, the minimum value for this configuration should likely be 2 * the DNS GC job interval.

### PR DESCRIPTION
**Pull request by vipul-21**
Configuring the FQDN GC interval based on the config provided by the user. This would determine how long that connections should be idle for after the DNS TTL expires.

Fixes: #13687

### CODE REVIEW

File: pkg/fqdn/config.go
Comment (pippolo84) [inline pkg/fqdn/config.go]: Typo: trailing `after DNS TTL expires.`

File: pkg/fqdn/name_manager_gc.go
Comment (pippolo84) [inline pkg/fqdn/name_manager_gc.go]: What we are actually doing here is giving the user the ability to customize the multiplier used to set the TTL.

Following the referenced issue, I think we should instead setting `int(2 * DNSGCJobInterval.Seconds())` as a default value and then overwrite it with `n.config.ActiveConnectionsTTL` if the user given value is equal or greater (maybe logging something in case it isn't, to signal that the GC will use the default value).

File: pkg/fqdn/name_manager_gc.go
Comment (pippolo84) [inline pkg/fqdn/name_manager_gc.go]: Please refrain from this style-only changes, they make it more difficult to focus on the core changes of the patch (during review and after the merge too, in case of a bisect) and when backporting it to previous stable versions.
If the current version is ok for the active linters and you don't need to change those lines, it's better to leave them as is.

File: pkg/option/config.go
Comment (pippolo84) [inline pkg/option/config.go]: Style-only change here too.

File: pkg/option/config.go
Comment (pippolo84) [inline pkg/option/config.go]: Style-only change here too.

File: pkg/option/config.go
Comment (pippolo84) [inline pkg/option/config.go]: Style-only change here too.

File: pkg/option/config.go
Comment (pippolo84) [inline pkg/option/config.go]: Style-only change here too.

File: pkg/option/config.go
Comment (pippolo84) [inline pkg/option/config.go]: Style-only change here too.

File: pkg/option/config.go
Comment (pippolo84) [inline pkg/option/config.go]: Style-only change here too.

File: pkg/option/config.go
Comment (pippolo84) [inline pkg/option/config.go]: Style-only change here too.

**Review by pippolo84 — CHANGES_REQUESTED** [review]
Thanks for the PR :pray: 
I think we should use the user given value in another way. Also, please be careful with those style only changes, they might make the patch harder to read without any additional benefit.

File: pkg/fqdn/name_manager_gc.go
Comment (joestringer) [reply pkg/fqdn/name_manager_gc.go]: Yes, for sure if the configuration is a TTL then it should be typed as a time and not a multiplier. I doubt that we really want to expose this configuration knob actually as a multiplier, that just makes things even more confusing for the user.

**Comment by joestringer** [comment]
Two more points here: How do you configure this? How did you test this?

File: pkg/fqdn/name_manager_gc.go
Comment (joestringer) [reply pkg/fqdn/name_manager_gc.go]: Agree, any stylistic changes should be paired with a linter that ensures everyone writes code the same way. The current Go standard linters/formatters don't reformat these lines, so I don't think it's a good idea to shuffle these lines around.

**Comment by jibi** [comment]
turning into a draft while the first round of comments get addressed

**Comment by vipul-21** [comment]
Update the PR with the suggested changes. (@pippolo84 Want to blame my editor for the styles changes. Sorry about that)
@joestringer For the testing, I configured it in the code base with changing default values. I checked the cache list in the cilium pod if the dns entry is still there after the TTL has expired. (until active connection ttl). Is there a way I can test it better or any other approach that is recommended ?

Also, based on the change I think we need to add the Documentation for the flag too. I haven't added that will wait until I get feedback on the new changes and how I can test it.

File: pkg/defaults/defaults.go
Comment (pippolo84) [inline pkg/defaults/defaults.go]: One thing that I am concerned about is that this default is set as a constant, without any reference to the `DNSGCJobInterval` value. But the CT GC and the Name Manager GC work together as per the comment in `pkg/fqdn/name_manager_gc.go`:

```
// activeConnections holds DNSName -> single IP entries that have been
// marked active by the CT GC. Since we expire in this controller, we
// give these entries 2 cycles of TTL to allow for timing mismatches
// with the CT GC.
```

So:

1) is it safe to set `ActiveConnectionsTTL` to a value lower than `2 * DNSGCJobInterval`? My understanding is that this can cause unexpected evictions of <name, IP> entries which should be kept alive because of an existing established connection.
2) what if in the future we change the `DNSGCJobInterval` value? I think it will be very easy to forget to adjust this related value too.

I think we can solve this keeping the default as `2 * DNSGCJobInterval`, thus moving the `DNSGCJobInterval` constant declaration here.

File: pkg/option/config.go
Comment (pippolo84) [inline pkg/option/config.go]: Again, here we are not checking that the user given value is actually >= `2*DNSGCJobInterval`, but I think we should.

**Review by pippolo84 — COMMENTED** [review]
Thanks for the further iteration. I left a suggestion about the consistency between CT GC and Name Manager GC default intervals.

cc @squeed

File: pkg/defaults/defaults.go
Comment (vipul-21) [reply pkg/defaults/defaults.go]: Thanks for the review, I agree with suggested changes and have updated the PR.

File: pkg/fqdn/name_manager_gc.go
Comment (vipul-21) [inline pkg/fqdn/name_manager_gc.go]: @pippolo84 Should we change this to `activeConnectionTTL` ?

Comment (joestringer) [reply pkg/fqdn/name_manager_gc.go]: Yeah that makes sense to me, that would match the equivalent `Update()` calls in `GC()`.

File: pkg/option/config.go
Comment (joestringer) [inline pkg/option/config.go]: If you add the flag to the daemon, then the daemon will already print out the configuration for this flag in a standardized way, so then there would be no need to print it here.
```suggestion
```

File: pkg/option/config.go
Comment (joestringer) [inline pkg/option/config.go:3231]: I'm not sure whether `vp.IsSet()` is the best option here. Consider the case where the user *doesn't* configure `ToFQDNsActiveConnectionsTTL` but *does* configure `DNSGCJobInterval`. The 2x multiplier check here could be invalidated due to the _other_ configuration.

Maybe this whole switch statement could be simplified just to a check of `if vp.GetInt(ToFQDNsActiveConnectionsTTL) < int(2*defaults.DNSGCJobInterval) { ...}` with the override to ensure it is double the GC Job interval?

File: pkg/option/config.go
Comment (joestringer) [inline pkg/option/config.go]: For the user-facing flag, I think it would help to define not just that the connection "should be idle for X long", but what the agent will do with entries based on this timer.

For instance:
```suggestion
	// ToFQDNsActiveConnectionsTTL is the time after which idle DNS entries will be garbage collected, following TTL expiry
```

(Though strictly speaking I don't think there's any bounding to say how quickly those entries _will_ be expired after this timeout, only that this sets a minimum period after which the timeout will occur).

File: daemon/cmd/fqdn.go
Comment (joestringer) [inline daemon/cmd/fqdn.go:1]: > For the testing, I configured it in the code base with changing default values. I checked the cache list in the cilium pod if the dns entry is still there after the TTL has expired. (until active connection ttl). Is there a way I can test it better or any other approach that is recommended ?

Broadly what I would hope is that `pkg/fqdn/name_manager_test.go` had some tests that validate the functionality that you're trying to achieve. For instance maybe this sets a new floor for the expiry time for active entries.

I was a bit hesitant to suggest a way to test this functionality because I couldn't quite tell what behaviour you are looking for when allowing this option to be configured; Is there a specific reason that you want to be able to tweak this setting, or is it more just that it is a tunable knob that is part of the GC algorithm and therefore it would be useful to be able to tweak the setting in case the GC is unresponsive?

**Review by joestringer — CHANGES_REQUESTED** [review]

File: pkg/option/config.go
Comment (vipul-21) [reply pkg/option/config.go:3231]: I don't understand the scenario, how can the user specify `DNSGCJobInterval` ? I thought we only allow user to specify `ToFQDNsActiveConnectionsTTL`

File: daemon/cmd/fqdn.go
Comment (vipul-21) [reply daemon/cmd/fqdn.go:1]: There isn't a specific reason for it. It more of this:
> is it more just that it is a tunable knob that is part of the GC algorithm and therefore it would be useful to be able to tweak the setting in case the GC is unresponsive

Comment (vipul-21) [reply daemon/cmd/fqdn.go:1]: Let me see if I can add that test case you are talking about.

File: pkg/option/config.go
Comment (joestringer) [reply pkg/option/config.go:3231]: Oh, you're right, I don't know why I thought that was configurable. Usually most things are configurable even if they're hidden.

I guess the intent behind the switch statement is to specifically identify the case where a user has specified an invalid setting, and override + log in that case? That's reasonable.

File: pkg/option/config.go
Comment (joestringer) [inline pkg/option/config.go]: Couple of things here:
- The third parameter is described in the string as a "user value" but the parameter is a hardcoded default
- Standard style in Cilium is to prefer structured logging and avoid formatting variations on logging functions like `Infof`, `Warningf`, `Errorf`. We may make exceptions to this in the case of hardcoded values (most commonly under `daemon/cmd/`), but when some fields of the log may be variable, they should be structured:

```suggestion
			log.WithFields(logrus.Fields{
				logfields.Key: ToFQDNsActiveConnectionsTTL,
				logfields.Value: defaults.ToFQDNsActiveConnectionsTTL,
				logfields.InvalidValue: vp.GetInt(ToFQDNsActiveConnectionsTTL),
			}).Info("Invalid configuration value for key, using default value")
```

**Review by joestringer — COMMENTED** [review]
Looks like this still needs to be exposed to the user via `daemon/cmd` and needs a rebase against the latest `main` branch. One more minor comment below.

File: pkg/option/config.go
Comment (joestringer) [reply pkg/option/config.go]: ^ We try to standardize the way that log fields are exposed to the user by defining them in the `logfields` package. `InvalidValue` isn't a logfield in that package yet, but I think the above would be a reasonable addition for this sort of case. Previously I think we may have just open-coded the string like `"invalidValue"` or `"oldValue"` but the whole idea with the `logfields` package is to ensure that logging fields are consistent over time.

File: daemon/cmd/fqdn.go
Comment (vipul-21) [reply daemon/cmd/fqdn.go:1]: I tried adding the test case, but since there isn't an active pathway for testing the GC. It seems quite difficult to check if the activeConnectionTTL has been GC'ed by the nameManager. 
I have added the test case where the TTL is being checked or acknowledged by the GC.
Not sure if the GC case blocks the PR or not.

Comment (vipul-21) [reply daemon/cmd/fqdn.go:1]: cc: @joestringer

File: pkg/fqdn/name_manager_test.go
Comment (joestringer) [inline pkg/fqdn/name_manager_test.go]: nit: If the times are seconds, then they should probably be declared as so. This will avoid misuse later on, and save some casting:

```suggestion
		activeConnectionTTL = 2 * time.Second
		nameManager         = NewNameManager(Config{
			MinTTL:               1 * time.Second,
```

File: pkg/fqdn/name_manager_test.go
Comment (joestringer) [inline pkg/fqdn/name_manager_test.go:192]: This test is not doing anything. The `nameManager.cache.forward` slice is empty.

**Review by joestringer — CHANGES_REQUESTED** [review]

File: pkg/fqdn/name_manager_test.go
Comment (vipul-21) [reply pkg/fqdn/name_manager_test.go:192]: Why is that ? I am seeing the length of cache.forward being 1 after adding a break point.
image

Comment (joestringer) [reply pkg/fqdn/name_manager_test.go:192]: I pulled the PR fresh and it looks like I was mistaken, sorry for the noise. The NameManager cache starts empty, but since the endpoint has a pointer to a list of zombies where the zombie's "active" time is set to the start of the test, the GC call then pulls those zombies into the NameManager with the newly introduced active connection timeout.

File: pkg/fqdn/name_manager_test.go
Comment (joestringer) [inline pkg/fqdn/name_manager_test.go:186]: There's still something I don't understand about the use of time here. Specifically, `time.Duration(activeConnectionTTL)` is getting multiplied by `time.Second`, but `activeConnectionTTL` is defined as `int(2*time.Second)`. Duration is measured in nanoseconds, so `activeConnectionsTTL` would be 2e9, then the value of the parameter to `Add()` here would be 2e18 :thinking: . How is this then leading to the correct value for the expiration offset?

File: daemon/cmd/daemon_main.go
Comment (joestringer) [inline daemon/cmd/daemon_main.go:798]: Please specify the units for the flag (presumably seconds).

File: pkg/fqdn/config.go
Comment (joestringer) [inline pkg/fqdn/config.go:24]: What are the intended units for this integer value?

I think that we should probably be trying to store & use native time types for these configurations rather than storing integers and doing the type conversions on the fly. Unless there's another reason that makes this awkward, what do you think about specifying it with the following?

```suggestion
	// Active Connections TTL is the time used by the GC to determines how long that
	// connections should be idle for after the DNS TTL expires.
	ActiveConnectionsTTL time.Duration
```

Note that if we _do_ specify this with a native time type then we'll need to be a little more careful in how this is initialized from the user-provided argument, for instance if the flag is in seconds then multiply the user configuration by `time.Second`.

(In theory the same feedback should apply to `MinTTL` too, but that code is already there so I don't expect you to change it in this PR since it's unrelated. We can always iterate on that separately/after).

File: pkg/fqdn/name_manager_test.go
Comment (joestringer) [reply pkg/fqdn/name_manager_test.go:186]: It'd be nice if we can drop the casts from this code to simplify (may depend on changing some units of some fields).

Nit: `time.Now()` here should probably also rely on the same `now` value pulled at the start of the test rather than fetching a fresh "now" value. It's unlikely, but if there is a significant scheduling delay between `now` fetched at the start of the test and used to calculate timers vs. `time.Now()` at this point in the test, then it could cause the test to become unreliable due to reliance on wall clock + expecting the code between to take zero time.

File: pkg/fqdn/config.go
Comment (vipul-21) [reply pkg/fqdn/config.go:24]: I think specifying the `time.Duration` make sense to me as well. Need to check if there are any cases where this might cause problems

File: pkg/fqdn/name_manager_test.go
Comment (vipul-21) [reply pkg/fqdn/name_manager_test.go:186]: I see. You're right. If I add a debug point after `time.now()`, the test case won't pass because of the delay.
Let me use the same now().

Also, let me add a brief description here about the test case I am trying to achieve and what it is doing. Will update it this week.

**Pull request closed without merging**
